src/shadowbox/server/server_access_key.ts

```diff
@@ -90,73 +123,162 @@ export class ServerAccessKeyRepository implements AccessKeyRepository {
+      const accessKey = this.accessKeys[ai];
       if (accessKey.id === id) {
-        this.portProvider.freePort(accessKey.port);
-        this.keyConfig.data().accessKeys.splice(ai, 1);
-        this.keyConfig.write();
+        this.portProvider.freePort(accessKey.proxyParams.portNumber);
+        this.accessKeys.splice(ai, 1);
+        this.saveAccessKeys();
         this.updateServer();
         return true;
       }
     }
     return false;
   }
 
-  listAccessKeys(): IterableIterator<AccessKey> {
-    return this.keyConfig.data().accessKeys.map(
-        accessKeyJson => makeAccessKey(this.proxyHostname, accessKeyJson))[Symbol.iterator]();
+  listAccessKeys(): AccessKey[] {
+    // Deep copy the access key array to prevent returning references to keys or their properties.
+    return this.accessKeys.map(key => (key instanceof Object) ? {...key} : key);
   }
 
   renameAccessKey(id: AccessKeyId, name: string): boolean {
-    const accessKeyJson = this.getAccessKey(id);
-    if (!accessKeyJson) {
+    const accessKey = this.getAccessKey(id);
+    if (!accessKey) {
       return false;
     }
-    accessKeyJson.name = name;
+    accessKey.name = name;
     try {
-      this.keyConfig.write();
+      this.saveAccessKeys();
+    } catch (error) {
+      return false;
+    }
+    return true;
+  }
+
+  async setAccessKeyQuota(id: AccessKeyId, quota: AccessKeyQuota): Promise<boolean> {
+    if (!quota || quota.quotaBytes < 1 || quota.windowHours < 1) {
+      return false;
+    }
+    const accessKey = this.getAccessKey(id);
+    if (!accessKey) {
+      return false;
+    }
+    accessKey.quota = quota;
+    try {
+      this.saveAccessKeys();
+      const quotaStausChanged = await this.updateAccessKeyQuotaStaus(accessKey);
+      if (quotaStausChanged) {
+        // Reflect the access key quota status if it changed with the new quota.
+        await this.updateServer();
+      }
+    } catch (error) {
+      return false;
+    }
+    return true;
+  }
+
+  async removeAccessKeyQuota(id: AccessKeyId): Promise<boolean> {
+    const accessKey = this.getAccessKey(id);
+    if (!accessKey) {
+      return false;
+    }
+    const quotaStausChanged = accessKey.isOverQuota;
+    accessKey.quota = undefined;
+    accessKey.isOverQuota = false;
+    try {
+      this.saveAccessKeys();
+      if (quotaStausChanged) {
+        await this.updateServer();
+      }
     } catch (error) {
       return false;
     }
     return true;
   }
 
   getMetricsId(id: AccessKeyId): AccessKeyMetricsId|undefined {
-    const accessKeyJson = this.getAccessKey(id);
-    return accessKeyJson ? accessKeyJson.metricsId : undefined;
+    const accessKey = this.getAccessKey(id);
+    return accessKey ? accessKey.metricsId : undefined;
+  }
+
+  // Compares access key usage with collected metrics, marking them as under or over quota.
+  async enforceAccessKeyQuotas() {
+    let quotaStatusChanged = false;
+    for (const accessKey of this.accessKeys) {
```

Comment:
I wonder if this might be an accidental pessimization to query prometheus a bunch of times for each key as opposed to just calling updateServer() and being okay if nothing changed.

Reply:
We only query Prometheus once for each key that has a quota. I don't think the extra measure to update the server only when necessary overly complicates the code, but I would be okay with calling `updateServer` unconditionally if you disagree.

Please let me know if I misunderstood your point.